Add test for register configuration for modbus switch

janiversen opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add test cases for modbus switch register configuration.

Solve a couple of small bugs in the switch register implementation.



## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x ] The code change is tested and works locally.
- [x ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x ] There is no commented out code in this PR.
- [x ] I have followed the [development checklist][dev-checklist]
- [ x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

janiversen committed:
Add register tests for modbus switch.

Copy the coil tests to work for register tests.

janiversen committed:
check for false config in modbus tests.

add call to pytest.fail() if device cannot be defined.

janiversen committed:
correct bit error in register read.

Only LSB is to be considered for checking ON/OFF

janiversen committed:
allow test coverage to control switch.py

With both register and coil test the coverage is a log higher than 50%

janiversen commented:
@vzahradnik can you please check the switch with your configuration, with this PR you can see that the problem with your tests was not that "state" do not exist, but the config was not complete.

janiversen commented:
hmmm I need to add test for state_on / state_off (this might affect the switch production code). And also solve the 3.8 problem.

vzahradnik commented:
I'll check tomorrow. Thanks!

vzahradnik reviewed (approved):
It looks good, I will try your updated tests on Modbus Light.

vzahradnik commented on homeassistant/components/modbus/switch.py:
```diff
@@ -260,7 +260,7 @@ def _read_register(self) -> Optional[int]:
         self._available = True
 
-        return int(result.registers[0])
+        return bool(result.bits[0] & 1)
```
@janiversen I am getting the following error:
```bash
AttributeError: 'ReadHoldingRegistersResponse' object has no attribute 'bits'
2020-10-30 13:38:30 ERROR (MainThread) [homeassistant] Error doing job: Future exception was never retrieved
```

I am using my test branch for `Modbus Light`, where I copied your changed files. Is there anything else, which I am missing here? Perhaps a updated `pymodbus` library?

vzahradnik replied:
I rebased my code to latest `dev`. Same error.

janiversen replied:
My bad I did not test with python 3.8, I have updated the PR (and added a couple more tests).

vzahradnik replied:
Can you tell what has changed in python 3.8? Also, shouldn't we still try to convert the value to bool?

Something like:
```python
return bool(result.registers[0] & 1)
```

Let me test the coil switch, too...

janiversen replied:
I wanted to change it to a bool, BUT I ran into problems since you can define state_on and state_off, therefore it is no longer simply the LSB.

I am not sure but the .bits works for coil but not register and both work in python 3.7

vzahradnik replied:
Did you check the coil switch on 3.8? I'm just asking so that we find a regression if there's any.

janiversen replied:
No I did not, but github runs the test suite on 3.8 (which is how I saw the problem).

janiversen committed:
Correct minor bug in handling of register in modbus switch

janiversen committed:
Add state_on/state_off test for modbus switch

ctalkington committed:
restore modbus switch exclusion from coverage

switch file coverage is only at 72%, ideally we want either 80%+ coverage on the file or 90%+ on the component itself which is only about 83% currently with this file included. this is so we keep the overall project coverage in 90%+.

ctalkington reviewed (approved):
just waiting on tests to pass with my minor tweak
